samples/azure/performance-tuning-asb/sample.md

```diff
@@ -0,0 +1,133 @@
+`Receiver` receives `SomeMessage` and computes every second how many it was able to receive in that timeframe
+
+Again there are different permutations of settings that have impact on the combines receive throughput.
+
+## SenderReceiver
+
+`SenderReceiver` is a variation on `Receiver` that does not only receive messages, but it also sends messages to a destination queue.
+
+## Running the sample
+
+ * Run the sender standalone to test send performance, but also to fill up the receive queue.
+ * Run the receiver standalone to test receive performance, if the queue empties, just run the sender again.
+ * Run the sender receiver standalone
+ 
+## Variations
+ 
+### Slow Sender
+
+In the slow sender sample, everything is done wrong:
+ * Only 1 factory is defined, this means that only 1 tcp connection is maintained with the broker.
+ * Client side batching of the ASB SDK is turned off, so the SDK will send message by message.
+
+Snippet: slow-send-config
+
+And each send is awaited individually, another cause that would prevent batching (even if batching client side batching were on).
+
+Snippet: slow-send  
+
+### Fast Sender
+
+The fast sender is configured much better:
+ * Multiple connections are established by configuring multiple messaging factories.
+ * Each factory is matched with a sender object (keep these equal so that there is 1 sender per factory)
+ * Client side batching is on, allowing the ASB SDK to send many messages at once
+
+Snippet: fast-send-config
+
+Individual sends are not awaited, instead the entire batch is awaited. Code execution will continue when all batches have been sent.
+
+Snippet: fast-send
+
+### Slow Atomic Receiver
+
+In the slow atomic receiver sample, everything is configured wrong:
+ * Only 1 factory is defined, this means that only 1 tcp connection is maintained with the broker.
+ * Prefetching is turned off, meaning that the receiver will fetch message by message
+ * No concurrent operations allowed, messages are processed one by one.
+
+Snippet: slow-atomic-receiver-config
+
+### Fast Atomic Receiver
+
+The fast atomic receiver is configured much better
+ * Multiple connections are established with the broker.
+ * Each connection is matched with exactly 1 receive client.
+ * Ensures that at least 16 connections are established when using partitioned queues, as each partitioned queue consists of 16 partitions. Depending on available bandwidth configure more until the network saturates.
+ * TransactionMode AtomicSendsWithReceive will execute complete operations in a serializable transaction, which means one by one. Therefore it is not beneficial to allow high per receiver concurrency. A value of 2 allows one receive to start while another is allowed to complete. The more time execution takes the higher this value can be, but never use really high values here.
+ * PrefetchCount is set to a relatively low number, as per receiver concurrency and throughput is low it makes no sense to load many messages at once.
```

Comment:
> PrefetchCount is set to a relatively low number, as per receiver concurrency and throughput is low it makes no sense to load many messages at once.

PrefetchCount is set to a relatively low number, since receiver concurrency and throughput are also low.